File: examples/python/inflation_index_example.py
#!/usr/bin/env python3
"""
Example: Using the new InflationIndex API for CPI/RPI calculations.

This example demonstrates how to:
1. Create inflation indices from historical observations
2. Use different interpolation methods
3. Apply lag policies for bond calculations
4. Calculate index ratios for inflation adjustments
"""

from typing import List, Tuple
from finstack import Date, Currency
from finstack.market_data import (
    InflationIndex, 
    InflationIndexBuilder,
    InflationInterpolation, 
    InflationLag
)


def create_us_cpi() -> InflationIndex:
    """Create a US CPI index with historical data."""
    # Historical CPI observations (monthly)
    observations: List[Tuple[Date, float]] = [
        (Date(2023, 1, 31), 299.170),
        (Date(2023, 2, 28), 300.840),
        (Date(2023, 3, 31), 301.836),
        (Date(2023, 4, 30), 302.918),
        (Date(2023, 5, 31), 303.294),
        (Date(2023, 6, 30), 303.841),
        (Date(2023, 7, 31), 304.348),
        (Date(2023, 8, 31), 305.537),
        (Date(2023, 9, 30), 306.269),
        (Date(2023, 10, 31), 306.139),
        (Date(2023, 11, 30), 306.060),
        (Date(2023, 12, 31), 306.746),
    ]
    
    # Create index with step interpolation (standard for CPI)
    cpi = InflationIndex(
        "US-CPI-U",
        observations,
        Currency("USD"),
        interpolation=InflationInterpolation.STEP,
        lag=InflationLag.months(3)  # 3-month lag for US TIPS
    )
    
    return cpi


def create_uk_rpi_builder() -> InflationIndex:
    """Create UK RPI using the builder pattern."""
    builder = InflationIndexBuilder("UK-RPI", Currency("GBP"))
    
    # Add observations incrementally
    builder.add_observation(Date(2023, 1, 31), 348.7)
    builder.add_observation(Date(2023, 2, 28), 351.2)
    builder.add_observation(Date(2023, 3, 31), 352.6)
    builder.add_observation(Date(2023, 4, 30), 354.0)
    builder.add_observation(Date(2023, 5, 31), 355.3)
    builder.add_observation(Date(2023, 6, 30), 356.2)
    
    # Configure interpolation and lag
    builder.with_interpolation(InflationInterpolation.LINEAR)
    builder.with_lag(InflationLag.months(2))  # 2-month lag for UK ILBs
    
    return builder.build()


def demonstrate_interpolation():
    """Compare different interpolation methods."""
    print("\n=== Interpolation Methods Comparison ===")
    
    observations = [
        (Date(2023, 1, 31), 300.0),
        (Date(2023, 2, 28), 303.0),
        (Date(2023, 3, 31), 306.0),
    ]
    
    # Step interpolation
    step_index = InflationIndex(
        "CPI-STEP",
        observations,
        Currency("USD"),
        interpolation=InflationInterpolation.STEP
    )
    
    # Linear interpolation
    linear_index = InflationIndex(
        "CPI-LINEAR",
        observations,
        Currency("USD"),
        interpolation=InflationInterpolation.LINEAR
    )
    
    # Test date between observations
    test_date = Date(2023, 2, 15)
    
    print(f"Test date: {test_date}")
    print(f"Step interpolation value: {step_index.value_on(test_date):.2f}")
    print(f"Linear interpolation value: {linear_index.value_on(test_date):.2f}")


def calculate_tips_inflation_adjustment(cpi: InflationIndex):
    """Calculate inflation adjustment for a TIPS bond."""
    print("\n=== TIPS Inflation Adjustment ===")
    
    # Bond details
    issue_date = Date(2023, 1, 15)
    settlement_date = Date(2023, 10, 15)
    
    # Get index values (with 3-month lag built in)
    base_index = cpi.value_on(issue_date)
    settlement_index = cpi.value_on(settlement_date)
    
    # Calculate index ratio
    index_ratio = cpi.ratio(issue_date, settlement_date)
    
    print(f"Issue date: {issue_date}")
    print(f"Settlement date: {settlement_date}")
    print(f"Base index (with lag): {base_index:.3f}")
    print(f"Settlement index (with lag): {settlement_index:.3f}")
    print(f"Index ratio: {index_ratio:.6f}")
    
    # Example: $1,000,000 principal
    principal = 1_000_000
    adjusted_principal = principal * index_ratio
    inflation_adjustment = adjusted_principal - principal
    
    print(f"\nOriginal principal: ${principal:,.2f}")
    print(f"Adjusted principal: ${adjusted_principal:,.2f}")
    print(f"Inflation adjustment: ${inflation_adjustment:,.2f}")


def compare_lag_policies(observations: List[Tuple[Date, float]]):
    """Compare different lag policies."""
    print("\n=== Lag Policy Comparison ===")
    
    test_date = Date(2023, 6, 15)
    
    # No lag
    no_lag = InflationIndex(
        "CPI-NO-LAG",
        observations,
        Currency("USD"),
        lag=InflationLag.none()
    )
    
    # 2-month lag
    lag_2m = InflationIndex(
        "CPI-2M-LAG",
        observations,
        Currency("USD"),
        lag=InflationLag.months(2)
    )
    
    # 3-month lag
    lag_3m = InflationIndex(
        "CPI-3M-LAG",
        observations,
        Currency("USD"),
        lag=InflationLag.months(3)
    )
    
    # 90-day lag
    lag_90d = InflationIndex(
        "CPI-90D-LAG",
        observations,
        Currency("USD"),
        lag=InflationLag.days(90)
    )
    
    print(f"Test date: {test_date}")
    print(f"No lag: {no_lag.value_on(test_date):.3f}")
    print(f"2-month lag: {lag_2m.value_on(test_date):.3f}")
    print(f"3-month lag: {lag_3m.value_on(test_date):.3f}")
    print(f"90-day lag: {lag_90d.value_on(test_date):.3f}")


def calculate_inflation_rates(cpi: InflationIndex):
    """Calculate various inflation rates."""
    print("\n=== Inflation Rate Calculations ===")
    
    # Get date range
    date_range = cpi.date_range()
    if date_range:
        start, end = date_range
        print(f"Index data range: {start} to {end}")
    
    # Calculate quarterly inflation
    q1_start = Date(2023, 1, 1)
    q1_end = Date(2023, 3, 31)
    q1_ratio = cpi.ratio(q1_start, q1_end)
    q1_rate = (q1_ratio - 1.0) * 100
    
    print(f"\nQ1 2023 inflation: {q1_rate:.2f}%")
    
    # Calculate semi-annual inflation
    h1_start = Date(2023, 1, 1)
    h1_end = Date(2023, 6, 30)
    h1_ratio = cpi.ratio(h1_start, h1_end)
    h1_rate = (h1_ratio - 1.0) * 100
    
    print(f"H1 2023 inflation: {h1_rate:.2f}%")
    
    # Calculate year-to-date inflation
    ytd_start = Date(2023, 1, 1)
    ytd_end = Date(2023, 11, 30)
    ytd_ratio = cpi.ratio(ytd_start, ytd_end)
    ytd_rate = (ytd_ratio - 1.0) * 100
    
    print(f"YTD inflation (through Nov): {ytd_rate:.2f}%")
    
    # Annualized rate
    days = (ytd_end - ytd_start)
    annualized_rate = ((ytd_ratio ** (365 / days)) - 1.0) * 100
    print(f"Annualized inflation rate: {annualized_rate:.2f}%")


def main():
    """Main example execution."""
    print("=" * 60)
    print("InflationIndex Example - Historical CPI/RPI Calculations")
    print("=" * 60)
    
    # Create US CPI index
    us_cpi = create_us_cpi()
    print(f"\nCreated: {us_cpi}")
    print(f"  Currency: {us_cpi.currency}")
    print(f"  Interpolation: {us_cpi.interpolation}")
    print(f"  Lag policy: {us_cpi.lag}")
    
    # Create UK RPI using builder
    uk_rpi = create_uk_rpi_builder()
    print(f"\nCreated: {uk_rpi}")
    print(f"  Currency: {uk_rpi.currency}")
    print(f"  Interpolation: {uk_rpi.interpolation}")
    print(f"  Lag policy: {uk_rpi.lag}")
    
    # Demonstrate different features
    demonstrate_interpolation()
    calculate_tips_inflation_adjustment(us_cpi)
    
    # Use the US CPI observations for lag comparison
    us_observations = [
        (Date(2023, 1, 31), 299.170),
        (Date(2023, 2, 28), 300.840),
        (Date(2023, 3, 31), 301.836),
        (Date(2023, 4, 30), 302.918),
        (Date(2023, 5, 31), 303.294),
        (Date(2023, 6, 30), 303.841),
    ]
    compare_lag_policies(us_observations)
    
    calculate_inflation_rates(us_cpi)
    
    print("\n" + "=" * 60)
    print("Example completed successfully!")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: examples/rust/inflation_index_example.rs
#!/usr/bin/env rust-script
//! Example demonstrating the new Polars-based InflationIndex API
//!
//! ```cargo
//! [dependencies]
//! finstack-core = { path = "../finstack/core" }
//! time = "0.3"
//! polars = "0.49"
//! ```

use finstack_core::dates::{InflationIndex, InflationIndexBuilder, InflationInterpolation, InflationLag};
use finstack_core::{Currency, Date};
use time::Month;

fn main() -> finstack_core::Result<()> {
    println!("=== Polars-based InflationIndex Demo ===\n");

    // Helper function for dates
    let make_date = |year, month: u8, day| {
        Date::from_calendar_date(year, Month::try_from(month).unwrap(), day).unwrap()
    };

    // 1. Create a US CPI index with historical observations
    println!("1. Creating US CPI index with monthly observations:");
    let observations = vec![
        (make_date(2023, 1, 31), 299.170),
        (make_date(2023, 2, 28), 300.840),
        (make_date(2023, 3, 31), 301.836),
        (make_date(2023, 4, 30), 302.918),
        (make_date(2023, 5, 31), 303.294),
        (make_date(2023, 6, 30), 303.841),
        (make_date(2023, 7, 31), 304.348),
        (make_date(2023, 8, 31), 305.537),
        (make_date(2023, 9, 30), 306.269),
        (make_date(2023, 10, 31), 306.139),
        (make_date(2023, 11, 30), 306.060),
        (make_date(2023, 12, 31), 306.746),
    ];

    let cpi = InflationIndex::new("US-CPI-U", observations.clone(), Currency::USD)?;
    println!("  Created {} with {} observations", cpi.id, observations.len());
    
    let (start, end) = cpi.date_range()?;
    println!("  Date range: {} to {}", start, end);

    // 2. Test different interpolation methods
    println!("\n2. Comparing interpolation methods for mid-month value:");
    let test_date = make_date(2023, 6, 15);
    
    let step_value = cpi.value_on(test_date)?;
    println!("  Step (default): {:.3}", step_value);
    
    let linear_cpi = InflationIndex::new("US-CPI-U", observations.clone(), Currency::USD)?
        .with_interpolation(InflationInterpolation::Linear);
    let linear_value = linear_cpi.value_on(test_date)?;
    println!("  Linear: {:.3}", linear_value);

    // 3. Calculate inflation rates
    println!("\n3. Calculating inflation rates:");
    let q1_start = make_date(2023, 1, 1);
    let q1_end = make_date(2023, 3, 31);
    let q1_ratio = cpi.ratio(q1_start, q1_end)?;
    println!("  Q1 2023: {:.2}%", (q1_ratio - 1.0) * 100.0);

    let h1_start = make_date(2023, 1, 1);
    let h1_end = make_date(2023, 6, 30);
    let h1_ratio = cpi.ratio(h1_start, h1_end)?;
    println!("  H1 2023: {:.2}%", (h1_ratio - 1.0) * 100.0);

    let ytd_start = make_date(2023, 1, 1);
    let ytd_end = make_date(2023, 12, 31);
    let ytd_ratio = cpi.ratio(ytd_start, ytd_end)?;
    println!("  Full year 2023: {:.2}%", (ytd_ratio - 1.0) * 100.0);

    // 4. Demonstrate lag policies (important for inflation-linked bonds)
    println!("\n4. Testing lag policies for TIPS/ILB calculations:");
    let settlement = make_date(2023, 10, 15);
    
    let no_lag = cpi.value_on(settlement)?;
    println!("  No lag: {:.3}", no_lag);
    
    let cpi_3m_lag = InflationIndex::new("US-CPI-U", observations.clone(), Currency::USD)?
        .with_lag(InflationLag::Months(3));
    let lag_3m = cpi_3m_lag.value_on(settlement)?;
    println!("  3-month lag: {:.3} (typical for US TIPS)", lag_3m);

    let cpi_2m_lag = InflationIndex::new("US-CPI-U", observations.clone(), Currency::USD)?
        .with_lag(InflationLag::Months(2));
    let lag_2m = cpi_2m_lag.value_on(settlement)?;
    println!("  2-month lag: {:.3} (typical for UK ILBs)", lag_2m);

    // 5. Using the builder pattern
    println!("\n5. Using the builder pattern for UK RPI:");
    let uk_rpi = InflationIndexBuilder::new("UK-RPI", Currency::GBP)
        .add_observation(make_date(2023, 1, 31), 348.7)
        .add_observation(make_date(2023, 2, 28), 351.2)
        .add_observation(make_date(2023, 3, 31), 352.6)
        .add_observation(make_date(2023, 4, 30), 354.0)
        .with_interpolation(InflationInterpolation::Linear)
        .with_lag(InflationLag::Months(2))
        .build()?;

    println!("  Created {}", uk_rpi.id);
    let uk_value = uk_rpi.value_on(make_date(2023, 6, 15))?;
    println!("  Value on 2023-06-15 with 2-month lag: {:.1}", uk_value);

    // 6. Access underlying Polars DataFrame for advanced operations
    println!("\n6. Leveraging Polars DataFrame capabilities:");
    let df = cpi.as_dataframe();
    println!("  DataFrame shape: {:?}", df.shape());
    println!("  Column names: {:?}", df.get_column_names());
    
    // You can now use all Polars operations on this DataFrame
    // e.g., df.filter(), df.group_by(), df.join(), etc.
    
    println!("\n✅ All tests completed successfully!");
    Ok(())
}
